proposals/lip-0055.md

```diff
@@ -105,16 +115,42 @@ The full processing of a block is organized as follows.
 14. **Result verification**: Block header properties, which require accessing the state store *after* all state transitions implied by the block have been executed, are verified.
 15. **Block storage**: The block is persisted into the database.
 16. **Peers notification**: Other peers in the P2P network are notified of the new block.
 
+
+### Block Generation Stages
+
+<img src="lip-0055/generation_stages.png" width="70%">
+
+*Figure 2: A schematic depiction of the various stages of a block generation. The steps performed in the consensus domain are indicated by blue boxes, while the steps performed in the state-machine domain are indicated by red boxes. Within the state-machine domain, the transactions processing is indicated by purple boxes. The network domain is indicated by green boxes.*
+
+In this section, we describe the various stages of a block generation.
+
+Similarly to the block processing, the block generation is split between the network, consensus, and state-machine domains.
+The full generation of a block is organized as follows.
+
+1. **Header creation**: Block header properties that require access to the state store *before* any state transitions implied by the block are executed are inserted in this stage.
+2. **Block initialization**: Block assets properties that require access to the state store *before* any state transitions implied by the block are executed are inserted in this stage.
+3. **Before transactions execution**: Each module can define protocol logic that is executed *before* the transactions contained in the block are processed.
+4. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. From this step until the end of the "before command execution" step, if an error occurs the transaction is invalid and it cannot be included in the block.
+5. **Command verification**: The command corresponding to the `moduleID`-`commandID` combination is verified.
+6. **Before command execution**: Each module can define protocol logic that is processed *before* the command has been executed. After this step is completed, the transaction is guaranteed to be included in the block payload.
+7. **Command execution**: The command corresponding to the `moduleID`-`commandID` combination is executed.
+8. **After command execution**: Each module can define protocol logic that is processed *after* the command has been executed.
+9. **After transactions execution**: Each module can define protocol logic that is executed *after* all the transactions contained in the block have been processed.
+9. **Block sealing**: Block assets properties that require access to the state store *after* any state transitions implied by the block are executed are inserted in this stage.
+11. **Result verification**: Block header properties, which require accessing the state store *after* all state transitions implied by the block have been executed, are inserted.
```

Comment:
> are inserted.

If these properties are inserted instead of verified, then stage should have a different name. For example "Block Header Completion". But it's still weird to add some properties to the block (header) after the "block sealing" stage. Maybe the order of stages is not correct, or we should also rename the "block sealing" stage.

Reply:
For block generation specifically, probably this is not `Result verification`.
Maybe we can say `Block header update` or `Enclose block header` (i cannot think of good name here)? Also, signing of the block header should happen here

Reply:
Note that in the diagram, this is also called result calculation which is inconsistent. I agree with @shuse2 that it would be good to use a different name here (block sealing seems more suitable for this hook, then for the previous one which is only about adding data to the asset).

Reply:
diagram updated

Reply:
All the new names look good from my side. We could resolve this comment.